docs/source/random_variable_design.rst

```diff
@@ -0,0 +1,203 @@
+Parameterize each ``RandomVariable`` in terms of other ``RandomVariable``\ s
+============================================================================
+
+If a ``RandomVariable`` ``Var`` needs to know some numerical parameter
+values ``a``, ``b``, etc to perform its ``sample()`` call, those values
+should be represented as constiuent ``RandomVariables`` that are
+attributes of ``Var`` (i.e. ``Var.a_rv``, ``Var.b_rv``) and are are
+``sample()``-d within ``Var.sample()`` (i.e. ``Var.sample()`` gets an
+``a`` value by calling ``self.a_rv.sample()`` and a ``b`` value by
+calling ``self.b_rv.sample()``).
```

Comment:
I think this might be too strong. Consider or generic observation processes, which have a mean argument. The mean is almost guaranteed to have been sampled by some other, more complex, RandomVariable that should not itself be wrapped by the observation process:

```python
        # Sampling from the latent process
        (
            post_initialization_latent_infections,
            *_,
        ) = self.latent_infections_rv.sample(
            Rt=Rt,
            gen_int=gen_int,
            I0=I0,
            **kwargs,
        )

        observed_infections, *_ = self.infection_obs_process_rv.sample(
            mu=post_initialization_latent_infections[padding:],
            obs=data_observed_infections,
            **kwargs,
        )
```

This document suggests that `post_initialization_latent_infections[padding:]` should be made into a deterministic variable and then passed to the observation process. That seems a bit too much.

Reply:
Concretely, this means allowing random variables to have a mix of parameters defined at construction (which must be `RandomVariable`s, possibly deterministic and parameters passed to the `sample()` call. I would support this, but would specify that such parameters should always be fixed / "observed" / "data" from the RV's point-of-view. This aligns with the `numpyro` pattern of passing `obs` arguments to `numpyro.sample` calls.

Reply:
I've encountered this too in #262. There's no silver bullet here. What if we need a `TimeArray` to be passed? I think the call to `RV.sample()` should be open to receive whatever (as it does now). Whatever could be a `RandomVariable`, a `scalar, an `array`, a `SampledArray` (aka `TimeArray`).

Reply:
We _could_ go stricter though, I do think it's worth thinking about the potential benefits and costs. Will add some thoughts to the document.

Reply:
> We could go stricter though, I do think it's worth thinking about the potential benefits and costs. Will add some thoughts to the document.

I agree with Damon & George's earlier sentiments on this but also do think we could go stricter in certain instances, based on what we know about the RVs parameters and potentially also the complexity of the RV / where it lies in the modeling process. Some additional examples might be helpful showcasing the verbosity of the maximally random approach in situations where it feels like "too much"; if these MSR examples end up seeming "not too bad", then I might be more convinced to operate in a maximally random way with MSR.

Reply:
In recent conversations, @dylanhmorris seem to have come to an agreement that RandomVariables should only have a single stochastic element, as a rule of thumb.